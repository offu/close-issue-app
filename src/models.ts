import { isArray, isString } from 'util'

/** Config for issue matching. */
export interface IssueConfig {
  /** content to match */
  content: Array<string>
}

/** Config for the issue bot */
export interface BotConfig {
  /** One BotConfig can include several IssueConfig for different issues */
  issueConfigs: Array<IssueConfig>
  /** Comment that will be sent if an issue is judged to be closed */
  comment: string
}

/**
 * check something if it is an IssueConfig
 */
export function isIssueConfig (item: any): item is IssueConfig {
  if (!('content' in item
  && isArray(item.content))) {
    return false
  }
  return item.content.every(isString)
}
/**
 * check something if it is a BotConfig
 */
export function isBotConfig (item: any): item is BotConfig {
  if (!('comment' in item
  && 'issueConfigs' in item
  && isString(item.comment)
  && isArray(item.issueConfigs))) {
    return false
  }
  return item.issueConfigs.every(isIssueConfig)
}
