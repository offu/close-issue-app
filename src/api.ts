import { Context } from 'probot'
import { Buffer } from 'buffer'

export async function closeIssue (context: Context) {
  const params = {
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name,
    issue_number: context.payload.issue.number
  }
  await context.github.issues.update({
    ...params,
    state: 'closed'
  })
}

export async function createLabelIfNotExist (context, label: string) {
  try {
    const l = await context.github.issues.getLabel({
      owner: context.payload.repository.owner.login,
      repo: context.payload.repository.name,
      name: label
    })
  } catch (e) {
    if (e.hasOwnProperty('status') && e.status === 404) {
      await context.github.issues.createLabel({
        owner: context.payload.repository.owner.login,
        repo: context.payload.repository.name,
        name: label,
        color: 'aa1414'
      })
    } else {
      throw e
    }
  }
}

export async function addLabel (context: Context, label: string) {
  await createLabelIfNotExist(context, label)
  await context.github.issues.addLabels({
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name,
    number: context.payload.issue.number,
    labels: [label]
  })
}

export async function createComment (context: Context, comment: string) {
  const params = {
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name,
    number: context.payload.issue.number
  }
  await context.github.issues.createComment({
    body: comment,
    ...params
  })
}

export async function getContent (context: Context, path: string) {
  const params = {
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name
  }
  const resp = await context.github.repos.getContents({
    path,
    ...params
  })
  return Buffer.from(resp.data[0].content, resp.data[0].encoding).toString('utf-8')
}
