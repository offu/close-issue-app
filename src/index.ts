interface ProBot {
  log: any
}

export = (robot: ProBot) => {
  robot.log('Yay, the app was loaded!')
}
