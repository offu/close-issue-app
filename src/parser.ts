import * as yaml from 'js-yaml'
import { BotConfig, isBotConfig } from './models'

/**
 * Parse the file according to the path into BotConfig
 */
export function parseConfig (content: string): BotConfig {
  const data = yaml.safeLoad(content)
  if (!isBotConfig(data)) {
    throw new Error('invalid config')
  } else {
    return data
  }
}

export function shouldClose (botConfig: BotConfig, content: string): boolean {
  return !botConfig.issueConfigs.some((issueConfig) => {
    return issueConfig.content.every((issueContent) => {
      if (botConfig.caseInsensitive) {
        return content.toLowerCase().includes(issueContent.toLowerCase())
      } else {
        return content.includes(issueContent)
      }
    })
  })
}
