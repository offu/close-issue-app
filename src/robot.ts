import { Robot } from 'probot'

export = (robot: Robot) => {
  robot.log('Yay, the app was loaded!')
}
