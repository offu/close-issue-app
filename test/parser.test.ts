import { parseConfig } from '../src/parser'
import * as path from 'path'

describe('parseConfig', () => {
  it('right example', () => {
    const exampleConfig = parseConfig(path.resolve(__dirname, '../example.config.yml'))
    expect(exampleConfig).toMatchSnapshot()
  })
  it('no issueConfigs', () => {
    expect(() => {
      parseConfig(path.resolve(__dirname, './wrong.config.yml'))
    }).toThrowError('invalid config')
  })
})
