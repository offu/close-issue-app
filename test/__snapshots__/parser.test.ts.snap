// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseConfig right example 1`] = `
Object {
  "comment": "test",
  "issueConfigs": Array [
    Object {
      "content": Array [
        "内容1",
        "内容2",
        "🐶",
      ],
    },
    Object {
      "content": Array [
        "🐱",
      ],
    },
  ],
}
`;
